cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/AppFabricServer.java

```diff
@@ -151,22 +152,36 @@ protected void startUp() throws Exception {
     LoggingContextAccessor.setLoggingContext(new ServiceLoggingContext(NamespaceId.SYSTEM.getNamespace(),
                                                                        Constants.Logging.COMPONENT_NAME,
                                                                        Constants.Service.APP_FABRIC_HTTP));
-    Futures.allAsList(
+    Future<List<State>> independentServicesFutures = Futures.allAsList(
       ImmutableList.of(
-        provisioningService.start(),
         applicationLifecycleService.start(),
         bootstrapService.start(),
         programRuntimeService.start(),
-        programNotificationSubscriberService.start(),
         programStopSubscriberService.start(),
         runRecordCorrectorService.start(),
         programRunStatusMonitorService.start(),
         coreSchedulerService.start(),
         eventPublishManager.start(),
         runRecordCounterService.start()
       )
+    );
+
+    //Services With Dependent services
+    Futures.allAsList(
+      ImmutableList.of(
+        provisioningService.start()
+      )
     ).get();
 
+    //Services that have a dependency on Successful start of one of the above services
+    Futures.allAsList(
+      ImmutableList.of(
+        programNotificationSubscriberService.start()
+      )
+    ).get();
+
+    independentServicesFutures.get();
```

Comment:
This approach results in independent services to start after ProvisioningService and ProgramNotificationSubscriberService getting started, which in turn prolongs appfabric start time. 

All we need to do is to make sure ProgramNotificationSubscriberService only starts after ProvisioningService. In other words, all services except ProgramNotificationSubscriberService can start in parallel. As soon as ProvisioningService is started, we can start ProgramNotificationSubscriberService.

I believe we can use CompletableFuture to enforce this invariant.

Reply:
We don't want to add ordering of service starts. If we ever need to start the service somewhere else, it will be very easy to do it improperly. Allowing this pattern also makes it very easy to get into a bad web of dependencies that becomes very difficult to manage.

Instead, there should be logic within the ProgramNotificationSubscriberService so that it doesn't use the relevant ProvisioningService methods if the service itself is not running. For example, see https://github.com/cdapio/cdap/blob/426a85da838b11c9f20a4ce5bc3f13270020998a/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java#L183.